connectors/sources/graphql.py

```diff
@@ -0,0 +1,519 @@
+                timeout=timeout,
+                raise_for_status=True,
+            )
+        else:
+            return aiohttp.ClientSession(
+                headers=self.headers,
+                timeout=timeout,
+                raise_for_status=True,
+            )
+
+    def extract_graphql_data_items(self, data):
+        """Returns sub objects from the response based on graphql_object_list
+
+        Args:
+            data (dict): data to extract
+
+        Yields:
+            dictionary/list: Documents from the response
+        """
+        for keys in self.graphql_object_list:
+            current_level_data = data
+            key_list = keys.split(".")
+
+            for key in key_list:
+                if isinstance(current_level_data, dict):
+                    if key not in current_level_data.keys():
+                        msg = (
+                            f"{key} not found in the response. Please check the query."
+                        )
+                        raise ConfigurableFieldValueError(msg)
+                    current_level_data = current_level_data.get(key, {})
+                else:
+                    msg = "Found list while processing GraphQL Objects List. Please check the query and provide path upto dictionary."
+                    raise ConfigurableFieldValueError(msg)
+
+            if isinstance(current_level_data, dict):
+                if current_level_data:
+                    yield current_level_data
+
+            if isinstance(current_level_data, list):
+                for doc in current_level_data:
+                    yield doc
+
+    def extract_pagination_info(self, data):
+        pagination_key_path = self.pagination_key.split(".")
+        for key in pagination_key_path:
+            if isinstance(data, dict):
+                if key not in data.keys():
+                    msg = f"Pagination key: {key} not found in the response. Please check the query."
+                    raise ConfigurableFieldValueError(msg)
+                data = data.get(key)
+            else:
+                msg = "Found list while processing Pagination key. Please check the query and provide path upto dictionary."
+                raise ConfigurableFieldValueError(msg)
+
+        if (
+            isinstance(data, dict)
+            and data.get("pageInfo")
+            and {"hasNextPage", "endCursor"}.issubset(set(data.get("pageInfo", {})))
+        ):
+            page_info = data.get("pageInfo")
+            has_next_page = page_info.get("hasNextPage")
+            end_cursor = page_info.get("endCursor")
+
+            return has_next_page, end_cursor, pagination_key_path[-1]
+        else:
+            msg = "Pagination is enabled but the query is missing 'pageInfo'. Please include 'pageInfo { hasNextPage endCursor }' in the query to support pagination."
+            raise ConfigurableFieldValueError(msg)
+
+    def validate_query(self, graphql_query, visitor):
```

Comment:
This only applies to paginated query? Can we rename this to `validate_paginated_query`?